addons/mail/tools/discuss.py

```diff
@@ -39,49 +41,67 @@ def get_sfu_key(env) -> str | None:
-            if not key in self.data:
-                self.data[key] = []
-            if isinstance(vals, dict):
-                vals = [vals]
-            if not isinstance(vals, list):
-                assert False, f"expected list for {key}: {vals}"
-            for val in vals:
-                if not isinstance(val, dict):
-                    assert False, f"expected dict for {key}: {val}"
-                for i in ids:
-                    if not val.get(i):
-                        assert False, f"missing key {i} in {key}: {val}"
-                match = filter(lambda record: all(record[i] == val[i] for i in ids), self.data[key])
-                if record := next(match, None):
-                    record.update(val)
-                else:
-                    self.data[key].append(val)
+    def add(self, data, values=None):
+        """Adds data to the store.
+        data can be a recordset, in which case the model must have a _to_store() method.
+        Or data can be a model name, in which case values must be a dict or list of dict.
+        Or data can be a dict, in which case it is considered as values for the Store model.
+        """
+        if isinstance(data, models.Model):
+            assert not values, f"expected empty values with recordset {data}: {values}"
+            records = data
+            model_name = target_model_by_model.get(records._name, records._name)
+            values = records._to_store(self)
+        elif isinstance(data, dict):
+            assert not values, f"expected empty values with dict {data}: {values}"
+            model_name = "Store"
+            values = data
+        else:
+            model_name = data
+        assert isinstance(model_name, str), f"expected str for model name: {model_name}: {values}"
+        # skip empty values
+        if not values:
+            return self
+        ids = ids_by_model[model_name]
+        # handle singleton: update in place
+        if len(ids) == 0:
+            assert isinstance(values, dict), f"expected dict for singleton {model_name}: {values}"
+            if not model_name in self.data:
+                self.data[model_name] = {}
+            self.data[model_name].update(values)
+            return self
+        # handle (multi) id(s): add or update existing
+        if not model_name in self.data:
+            self.data[model_name] = []
+        if isinstance(values, dict):
+            values = [values]
+        assert isinstance(values, list), f"expected list for {model_name}: {values}"
+        for vals in values:
+            assert isinstance(vals, dict), f"expected dict for {model_name}: {vals}"
+            for i in ids:
+                assert vals.get(i), f"missing id {i} in {model_name}: {vals}"
+            match = filter(lambda record: all(record[i] == vals[i] for i in ids), self.data[model_name])
+            if record := next(match, None):
+                record.update(vals)
+            else:
```

Comment:
`store.add()` twice on same record whose `many` field was `["ADD", {}]` will no combine the list of commands and instead just the latter list of commands will be present.

Not a huge problem for now, but just pointing out a limitation to be aware of.